### PR TITLE
fn: cleanup of docker timeouts and docker health check

### PR DESCRIPTION
**Pull request by skinowski**
Moving the timeout management of various docker operations
    to agent. This allows for finer control over what operation
    should use. For instance, for pause/unpause our tolerance
    is very low to avoid resource issues. For docker remove,
    the consequences of failure will lead to potential agent
    failure and therefore we wait up to 10 minute.
    For cookie create/prepare (which includes docker-pull)
    we cap this at 10 minutes by default.
    
    With new UDS/FDK contract, health check is now obsoleted
    as container advertise health using UDS availibility.

### CODE REVIEW

**Comment by skinowski** [comment]
This is a prep-work for https://github.com/fnproject/fn/pull/1288

File: api/agent/drivers/docker/docker.go [hunk 1]
```diff
@@ -577,36 +577,7 @@ func (drv *DockerDriver) startTask(ctx context.Context, container string) error
...
-
-		// if no health check for this image (""), or it's healthy, then stop waiting.
-		// state machine is "starting" -> "healthy" | "unhealthy"
-		if cont.State.Health.Status == "" || cont.State.Health.Status == "healthy" {
-			break
-		}
-
-		time.Sleep(100 * time.Millisecond) // avoid spin loop in case docker is actually fast
-	}
-	return nil
```
Comment (rdallman) [inline api/agent/drivers/docker/docker.go]: 👍 this makes sense with uds I think, good idea

File: api/agent/agent.go [hunk 2]
```diff
@@ -764,26 +770,32 @@ func (a *agent) runHot(ctx context.Context, call *call, tok ResourceToken, state
...
 	}
 
-	defer cookie.Close(ctx)
+	defer func() {
+		ctx, cancel := context.WithTimeout(common.BackgroundContext(ctx), removeTimeout)
+		cookie.Close(ctx)
+		cancel()
+	}()
 
 	err = a.driver.PrepareCookie(ctx, cookie)
```
Comment (rdallman) [inline api/agent/agent.go]: i forgot why create and prepare cookie got split up, maybe it was for cold, seems like we can put humpty dumpty back together again now (not implying that needs to get done here, just observing...)

File: api/agent/drivers/docker/docker.go [hunk 3]
```diff
@@ -276,6 +275,7 @@ func (drv *DockerDriver) PrepareCookie(ctx context.Context, c drivers.Cookie) er
 		return err
 	}
 
+	cookie.opts.Context = ctx
```
Comment (skinowski) [inline api/agent/drivers/docker/docker.go]: switching to correct ctx here, ctx passed in CreateCookie() should not be used in CreateContainer().

**Comment by rdallman** [comment]
yea, makes sense to move these to caller with these changes, couple notes

File: api/agent/drivers/docker/docker_client.go [hunk 4]
```diff
@@ -386,15 +383,9 @@ func (d *dockerWrap) PullImage(opts docker.PullImageOptions, auth docker.AuthCon
 }
 
 func (d *dockerWrap) RemoveContainer(opts docker.RemoveContainerOptions) (err error) {
-	// extract the span, but do not keep the context, since the enclosing context
-	// may be timed out, and we still want to remove the container. TODO in caller? who cares?
-	ctx := common.BackgroundContext(opts.Context)
-	ctx, closer := makeTracker(ctx, "docker_remove_container")
+	ctx, closer := makeTracker(opts.Context, "docker_remove_container")
```
Comment (skinowski) [inline api/agent/drivers/docker/docker_client.go]: moved this to caller as how to long to wait on this depends on higher level code.

File: api/agent/agent.go (hunk 2 above)
Comment (skinowski) [reply api/agent/agent.go]: cookie Iface doesn't make sense right now, also wait result. Another PR...  I've modified them here: https://github.com/fnproject/fn/pull/1279, which we can pick and carry into a separate PR later.

File: api/agent/agent.go [hunk 5]
```diff
@@ -27,6 +27,12 @@ import (
 	"os"
 )
 
+const (
+	removeTimeout = 1 * time.Minute // docker remove
+	prepTimeout   = 2 * time.Minute // docker create+pull
+	pauseTimeout  = 5 * time.Second // docker pause/unpause
```
[the remark this answers is not in the document]
Comment (skinowski) [reply api/agent/agent.go]: Increased that to 10 mins now. I'm keeping somewhat similar values as before.

**Comment by skinowski** [comment]
I think I should add another timeout here, docker-pull vs UDS/start. Blanket timeout for all of it seems too broad, and leaves too much time for UDS/start phase which will eat cpu/mem outside of call.Start/End.

File: api/agent/agent.go [hunk 6]
```diff
@@ -764,26 +768,29 @@ func (a *agent) runHot(ctx context.Context, call *call, tok ResourceToken, state
...
+	defer cancel()
+
 	cookie, err := a.driver.CreateCookie(ctx, container)
 	if err != nil {
 		call.slots.queueSlot(&hotSlot{done: make(chan struct{}), fatalErr: err})
 		return
 	}
 
-	defer cookie.Close(ctx)
+	// WARNING: we wait forever.
```
Comment (rdallman) [inline api/agent/agent.go]: retry in docker_client will give up at 10 tries and this should bail at that point, at least, currently I don't think this is true. so long as every call to the docker_client has a timeout in the context (in applicable cases, ie everything but this one?), we could remove the 10 retry cap, as the retry loop respects the context timeout/cancel as it is.

**Comment by rdallman** [comment]
> I think I should add another timeout here, docker-pull vs UDS/start. Blanket timeout for all of it seems too broad, and leaves too much time for UDS/start phase which will eat cpu/mem outside of call.Start/End.

that seems fine. original uds proposal had 15s or something, yea, they're on the clock for the uds thing.

**Pull request merged**